control/plugin/client/grpc.go

```diff
@@ -241,12 +281,142 @@ func (g *grpcClient) CollectMetrics(mts []core.Metric) ([]core.Metric, error) {
 	return metrics, nil
 }
 
+func (g *grpcClient) UpdateCollectedMetrics(mts []core.Metric) error {
+	if g.stream != nil {
+		arg := &rpc.CollectArg{
+			Metrics_Arg: &rpc.MetricsArg{Metrics: NewMetrics(mts)},
+		}
+		err := g.stream.Send(arg)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+func (g *grpcClient) UpdatePluginConfig(bytes []byte) error {
+	if g.stream != nil {
+		arg := &rpc.CollectArg{
+			Other: bytes,
+		}
+		err := g.stream.Send(arg)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+func (g *grpcClient) UpdateCollectDuration(maxCollectDuration time.Duration) error {
+	if g.stream != nil {
+		arg := &rpc.CollectArg{
+			MaxCollectDuration: maxCollectDuration.Nanoseconds(),
+		}
+		err := g.stream.Send(arg)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+func (g *grpcClient) UpdateMetricsBuffer(maxMetricsBuffer int64) error {
+	if g.stream != nil {
+		arg := &rpc.CollectArg{
+			MaxMetricsBuffer: maxMetricsBuffer,
+		}
+		err := g.stream.Send(arg)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (g *grpcClient) StreamMetrics(mts []core.Metric) (chan []core.Metric, chan error, error) {
+	arg := &rpc.CollectArg{
+		Metrics_Arg: &rpc.MetricsArg{Metrics: NewMetrics(mts)},
+	}
+	if len(mts) == 0 {
+		return nil, nil, errors.New("No metrics requested to stream")
+	}
+
+	mt := mts[0]
+	if cfg := mt.Config(); cfg != nil {
+		values := cfg.Table()
+		if values != nil {
+			maxCollectDuration, ok := values["MaxCollectDuration"]
+			if ok {
+				t, ok := maxCollectDuration.(*ctypes.ConfigValueInt)
+				if ok {
+					arg.MaxCollectDuration = int64(t.Value)
+				}
+			}
+			maxMetricsBuffer, ok := values["MaxMetricsBuffer"]
+			if ok {
+				t, ok := maxMetricsBuffer.(*ctypes.ConfigValueInt)
+				if ok {
+					arg.MaxMetricsBuffer = int64(t.Value)
+				}
+			}
+		}
+	}
+
+	s, err := g.streamCollector.StreamMetrics(context.Background())
+	if err != nil {
+		return nil, nil, err
+	}
+	err = s.Send(arg)
+	if err != nil {
+		return nil, nil, err
+	}
+	metricChan := make(chan []core.Metric)
+	errChan := make(chan error)
+	doneChan := make(chan struct{})
+	g.killChan = doneChan
+	g.stream = s
+	go g.handleInStream(metricChan, errChan)
+	return metricChan, errChan, nil
+}
+
+func (g *grpcClient) handleInStream(
+	metricChan chan []core.Metric,
+	errChan chan error) {
+	go func() {
+		done := false
+		for !done {
+			in, err := g.stream.Recv()
+			if err != nil {
+				errChan <- err
+				break
+			}
+			if in.Metrics_Reply != nil {
+				mts := ToCoreMetrics(in.Metrics_Reply.Metrics)
+				if len(mts) == 0 {
+					// skip empty metrics
+					continue
+				}
+				metricChan <- mts
+			} else if in.Error != nil {
+				e := errors.New(in.Error.Error)
+				errChan <- e
+			}
+		}
+	}()
+
+	<-g.killChan
+	errChan <- errors.New("connection broken")
```

Comment:
It's how we signal the scheduler to wait a bit before retrying to give the plugin time to be restarted. See [this](https://github.com/intelsdi-x/snap/pull/1526/files/330def915f9cdd002fe4aab095a9bbb5104eaac2#diff-77f163e381dd113e90ebd1bd76479b54R315)